hovor/actions/local_dialogue_action_simulated.py

```diff
@@ -0,0 +1,195 @@
+
+    def example_spacy_entity(self, category):
+        """
+        Since Spacy won't generate a given entity type from nothing,
+        I made this function which can generate a random example for
+        each category of named entities.
+        """
+        examples = {
+            "GPE": ['Kingston', 'Toronto'],
+            "NORP": ['The Liberal Party of Canada', 'The Conservative Party of Canada'],
+            "FAC": ['Toronto Pearson Airport', 'Golden Gate Bridge'],
+            "ORG": ['Amazon', "Queen's University"],
+            "PERSON": ['Ryan Reynolds', 'Harry Potter'],
+            "LOC": ['Africa', 'Mount Everest'],
+            "PRODUCT": ['Mazda 3', 'Smartphone'],
+            "EVENT": ['Olympic Games', 'Christmas'],
+            "WORK_OF_ART": ['Mona Lisa', 'The Starry Night'],
+            "LAW": ['The Civil Rights Act', 'The Patriot Act'],
+            "LANGUAGE": ['English', 'French'],
+            "DATE": ['16th Janurary 2017', 'November 12th'],
+            "TIME": ['ten minutes', 'four hours'],
+            "PERCENT": ['sixty percent', '20%'],
+            "MONEY": ['fifty dollars', '5 cents'],
+            "QUANTITY": ['Several kilometers', '60kg'],
+            "ORDINAL": ['9th', 'Ninth'],
+            "CARDINAL": ['2', 'Three'],
+        }
+        if category.upper() not in examples:
+            return None
+        return random.choice(examples[category.upper()])
+
+    def get_random_value_for_var(self, var_name):
+        cv = self.data['context_variables'][var_name]
+        if cv['type'] == 'enum':
+            return random.choice(cv['config'])
+        elif cv['type'] == 'json' and cv['config']['extraction']['method'] == 'spacy':
+            return self.example_spacy_entity(cv['config']['extraction']['config_method'])
+        else:
+            raise TypeError(
+                "Tried to fill in a non-enum or json var. Only enums and jsons are currently supported. ")
+
+    def fill_dollar_vars(self, s):
+        if '$' not in s:
+            return s
+        # finds the word starting with $ without the $
+        var_names = re.findall(r'\$(\w+)', s)
+        # get random values for these
+        random_var_values = [
+            self.get_random_value_for_var(n) for n in var_names]
+        # sub them into the string
+        for i in range(len(var_names)):
+            # only replace the first element incase there are two of same type vars
+            s = s.replace('$'+var_names[i], random_var_values[i], 1)
+        return s
+
+    def generate_response_by_required_entities(self):
+        """
+        A function that picks a random outcome group from self,
+        extracts the required entities in a response,
+        picks a random example of each intent,
+        filters these examples to ones that have the same entities as the required entities,
+        and returns a random choice from the possible examples as the response.
+
+        This works well for share_cuisine or share_outing_type,
```

Comment:
Also, do you care to simulate any entities that are "maybe" extracted? An easy way you could do this is to randomly change, add, or delete a letter in one of the "valid" entity options, i.e. "piza" instead of "pizza." I have a plugged-in spell checker in Rasa such that if the extracted entity is valid after a spell check, it gets a "maybe."

Reply:
Another thing - this is probably me misunderstanding the wording but if you are picking a random outcome you shouldn't have to then pick the intent as there should be one intent linked to every outcome already.

Finally, if you do end up taking "maybe"s into account you will have to look at the entity assignments in each outcome to determine the status of each entity. This shouldn't break anything but also note that multiple outcomes can have the same intent but with different entity assignments (mostly the case for `slot_fill` actions).

Reply:
Yes I agree I should remove the generate by required entities function -- I don't forsee a case where that would be a better approach.

Reply:
Also yes, this function was made before I started using data.json so it was trying to work around not having self.data defined. I will remove this function.

Reply:
The maybe's are an interesting idea. My first instinct is that I am primarily interested in when the chatbot does something weird upon normal inputs so I shouldn't simulate abnormal inputs, but there would definitely be value in looking at it. I will give it more thought, it might be out of the scope of my project.

Reply:
All good, if you think it is worth pursuing at some point, perhaps create an issue for it, but we can move past it for now :)